src/Core/Services/MultipleMutationInputValidator.cs

```diff
@@ -0,0 +1,608 @@
+                // All the nodes in the ListValueNode are parsed one by one.
+                listValueNode.GetNodes().ToList().ForEach(objectValueNodeInListValueNode => ValidateGraphQLValueNode(
+                    schema: schema,
+                    entityName: entityName,
+                    context: context,
+                    parameters: objectValueNodeInListValueNode,
+                    runtimeConfig: runtimeConfig,
+                    columnsDerivedFromParentEntity: columnsDerivedFromParentEntity,
+                    columnsToBeDerivedFromEntity: columnsToBeDerivedFromEntity,
+                    nestingLevel: nestingLevel,
+                    parentEntityName: parentEntityName,
+                    sqlMetadataProviderFactory: sqlMetadataProviderFactory));
+            }
+            else
+            {
+                throw new DataApiBuilderException(message: $"Unable to process input at level: {nestingLevel}",
+                    statusCode: HttpStatusCode.NotImplemented,
+                    subStatusCode: DataApiBuilderException.SubStatusCodes.NotSupported);
+            }
+        }
+
+        /// <summary>
+        /// Helper method to iterate over all the fields present in the input for the current field and
+        /// validate that the presence/absence of fields make logical sense for the multiple-create request.
+        /// </summary>
+        /// <param name="schemaObject">Input object type for the field.</param>
+        /// <param name="entityName">Current entity's name.</param>
+        /// <param name="context">Middleware Context.</param>
+        /// <param name="objectFieldNodes">List of ObjectFieldNodes for the the input field.</param>
+        /// <param name="runtimeConfig">Runtime config</param>
+        /// <param name="columnsDerivedFromSourceEntity">Set of columns in this entity whose values should be
+        /// derived from insertion in the source entity (i.e. source entity would have been the referenced entity).</param>
+        /// <param name="columnsToBeSuppliedToSourceEntity">Set of columns in this entity whose values are to be
+        /// derived from insertion in this entity or its subsequent referenced entities and returned to the source entity
+        /// so as to provide values for the corresponding referencing fields (i.e. source entity would have been the referencing entity).</param>
+        /// <param name="nestingLevel">Current depth of nesting in the multiple-create request. As we go down in the multiple mutation from the source
+        /// entity input to target entity input, the nesting level increases by 1. This helps us to throw meaningful exception messages to the user
+        /// as to at what level in the multiple mutation the exception occurred.</param>
+        /// <param name="parentEntityName">Parent entity's name.</param>
+        /// <param name="sqlMetadataProviderFactory">Metadata provider factory.</param>
+        private static void ValidateObjectFieldNodes(
```

Comment:
this was a huge function to read through/evaluate. Was there no way to break this up + have functions with smaller method signatures?

If this were a service instantiated at startup as non static and `scoped`(per request) service lifetime per https://learn.microsoft.com/dotnet/core/extensions/dependency-injection#scoped, you could use DI to resolve `RuntimeConfigProvider`/`RuntimeConfig`, `IMetadataProviderFactory`, reducing your parameter count everywhere in this file by 2 and allowing you to create short/low param count helper functions that don't need runtimeconfig/metadataprovider passed to them.

You may be able to further reduce parameter count (which increases readability) by introducing a `MultipleMutationRequestValidationContext` object which contains relevant/related properties such as
- `currentEntityName`
- `parentEntityName`
- `columnsDerivedFromSourceEntity`
- `columnsToBeSuppliedToSourceEntity`
- nestingLevel

but that may be for another PR. I just worry that in this may be tough for others to come in and understand without indepth exposure/understanding of our design discussions.

Reply:
and IMiddlewareContext is stored once per-request into a private var. That would require you to have a separate function that is recursed .

Reply:
> You may be able to further reduce parameter count (which increases readability) by introducing a MultipleMutationRequestValidationContext object which contains relevant/related properties such as

Introduced class `MultipleMutationEntityInputValidationContext`.

> this was a huge function to read through/evaluate. Was there no way to break this up + have functions with smaller method signatures?

Moved relationship processing in the method ValidateObjectFieldNodes() to a new method ProcessRelationshipField() to break up function into smaller ones.

> If this were a service instantiated at startup as non static and scoped(per request) 

For now, instantiating a multiple mutation input validator instance on a per request basis inside `SqlMutationEngine`. Not using DI right now as that might involve some more changes. Avoiding that at the moment.